qcodes/dataset/data_set.py

```diff
@@ -1333,26 +1395,29 @@ def _flush_data_to_database(self, block: bool = False) -> None:
                 argument has no effect if not using a background thread.
 
         """
+
         log.debug('Flushing to database')
+        writer_status = self._writer_status
         if len(self._results) > 0:
             try:
+
                 self.add_results(self._results)
-                if self._bg_writer.is_alive():
+                if writer_status.write_in_background:
                     log.debug(f"Succesfully enqueued result for write thread")
                 else:
                     log.debug(f'Successfully wrote result to disk')
                 self._results = []
             except Exception as e:
-                if self._bg_writer.is_alive():
+                if writer_status.write_in_background:
                     log.warning(f"Could not enqueue result; {e}")
                 else:
                     log.warning(f'Could not commit to database; {e}')
         else:
             log.debug('No results to flush')
 
-        if self._bg_writer.is_alive() and block:
+        if writer_status.write_in_background and block:
             log.debug(f"Waiting for write queue to empty.")
-            self._data_write_queue.join()
+            writer_status.data_write_queue.join()
```

Comment:
one question - perhaps i missed - from the DataSaver/RUnner side we never block on the flushing, even when exiting the measurement context manager (see here https://github.com/QCoDeS/Qcodes/blob/84648a52b79b3af7b48237f5062bd41302126fe8/qcodes/dataset/measurements.py#L491) - is that good? i was thinking that likely when exiting the outermost measurement context manager, the user really expects ALL the data to be successfully written? or is this an unreasonable ask?

Reply:
So this actually happens another way round. The ``__exit__`` method of the Runner context manager calls ``_ensure_dataset_written`` which will use the finalize message to ensure that a given dataset is written before the context manager can exit.

Reply:
aaah!! that's amazing ! :)